### PR TITLE
Refactoring SidebarWidget and sidebarUser

### PR DESCRIPTION
**Pull request by Rafi993**
## What type of PR is this? 

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Refactoring https://github.com/forem/forem/blob/master/app/javascript/sidebar-widget/SidebarWidget.jsx and
https://github.com/forem/forem/blob/master/app/javascript/sidebar-widget/sidebarUser.jsx
into functional component and simplifying the code

## Related Tickets & Documents
Closes https://github.com/forem/forem/issues/10948

## QA Instructions, Screenshots, Recordings
1. Go to url http://localhost:3000/t/discuss
2. Scroll down to see suggested users on the right panel and it should work without any change

## Added tests?

- [ ] yes
- [x] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.forem.com
- [ ] readme
- [x] no documentation needed

### CODE REVIEW

**Comment by msarit** [comment]
Hi @Rafi993! 👋🏾  Thanks for this PR. Think you could resolve the conflicts? Thanks so much.

**Comment by Rafi993** [comment]
> Hi @Rafi993! 👋🏾 Thanks for this PR. Think you could resolve the conflicts? Thanks so much.

I have fixed the merge conflicts now.

**Review by nickytonline — CHANGES_REQUESTED** [review]
Thanks for the PR Rafi! Just some change requests.

File: app/javascript/sidebar-widget/SidebarWidget.jsx [hunk 1]
```diff
@@ -1,34 +1,18 @@
...
-    this.setState({
-      tagInfo: JSON.parse(
+  useEffect(() => {
+    const tagInfo =
+      JSON.parse(
         document.getElementById('sidebarWidget__pack').dataset.tagInfo,
-      ),
-    });
-  }
+      ) || {};
```
Comment (nickytonline) [inline app/javascript/sidebar-widget/SidebarWidget.jsx:13]: Refactor (blocking): Setting a default value here does not seem necessary. We are always expecting `tagInfo` to be an object. If it isn't, Honeybader catches the unexpected error and reports it.

If we give it a a default value of `{}`, a valid request will be made that returns an empty array for the tag `undefined`.

File: app/javascript/sidebar-widget/SidebarWidget.jsx [hunk 2]
```diff
@@ -38,58 +22,57 @@ class SidebarWidget extends Component {
...
-    followBtn.innerText = updatedUser.following ? 'Follow' : 'Following';
-
-    const toggleFollowState = (newFollowState) => {
-      updatedUser.following = newFollowState === 'followed';
-      updatedSuggestedUsers[userIndex] = updatedUser;
-      this.setState({ suggestedUsers: updatedSuggestedUsers });
-    };
-    sendFollowUser(user, toggleFollowState);
-  }
+      const followBtn = document.getElementById(
```
Comment (nickytonline) [inline app/javascript/sidebar-widget/SidebarWidget.jsx:39]: Refactor (blocking): Using `useCallback` here and elsewhere in the PR seems like a premature optimization and it comes at a cost. See [When to useMemo and useCallback](https://kentcdodds.com/blog/usememo-and-usecallback).

**Review by nickytonline — CHANGES_REQUESTED** [review]
Just another small change request.

File: app/javascript/sidebar-widget/sidebarUser.jsx [hunk 3]
```diff
@@ -1,51 +1,43 @@
...
-    this.onClick = this.onClick.bind(this);
-  }
-
-  onClick() {
-    const { followUser, user } = this.props;
+const SidebarUser = ({ followUser, user }) => {
+  const onClick = useCallback(() => {
     followUser(user);
-  }
+  }, [followUser, user]);
```
Comment (aitchiss) [inline app/javascript/sidebar-widget/sidebarUser.jsx:9]: Refactor (blocking): I think in this case we don't want to employ `useCallback` as we won't get any performance boost from it ([Kent C Dodds has a great article about this](https://kentcdodds.com/blog/usememo-and-usecallback)) - instead we can keep it simple with `const onClick = () => followUser(user)`

**Review by aitchiss — CHANGES_REQUESTED** [review]
Looking good! Just one small change request 👍

**Review by aitchiss — APPROVED** [review]
LGTM - left one minor/nitpicky suggestion for you to take or leave 🙂

**Review by nickytonline — APPROVED** [review]
Rafi, looks good to me as well. Just the nitpick about the `index` prop as well like @aitchiss mentioned. I don't see it being used either.

**Comment by aitchiss** [comment]
Thanks for this one @Rafi993! ✨

**Pull request merged**